staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"fmt"
+	"sync"
+)
+
+// ready is a condition variable that blocks until is ready if is not stopped
+type ready struct {
+	ok      bool
+	stopped bool // if stopped is true ok must be false
+	c       *sync.Cond
+}
+
+func newReady() *ready {
+	return &ready{c: sync.NewCond(&sync.RWMutex{})}
+}
+
+// wait blocks until is ready or is stopped
+// it returns an error if is stopped.
+func (r *ready) wait() error {
+	r.c.L.Lock()
+	defer r.c.L.Unlock()
+	for !r.ok && !r.stopped {
+		r.c.Wait()
+	}
+	if r.stopped {
+		return fmt.Errorf("apiserver cacher is stopped")
+	}
+	return nil
+}
+
+// TODO: Make check() function more sophisticated, in particular
+// allow it to behave as "waitWithTimeout".
+func (r *ready) check() bool {
+	rwMutex := r.c.L.(*sync.RWMutex)
+	rwMutex.RLock()
+	defer rwMutex.RUnlock()
+	if r.stopped {
```

Comment:
when it starts both are false, I think that the tristate comment instead of two vars applies here, I can't remember now why I did this :) , maybe was easier to make it explicitly than thinking about it ...